### DISCUSSION TITLE
关于bee.filesystem的问题

### DISCUSSION THREAD

**Discussion by constly** [ejoy/ant#125] [General]
使用bee.filesystem的pairs函数遍历磁盘目录时，如果目录不存在游戏会卡死（类似死循环），调试后报错堆栈如下
![bug_06](https://github.com/constly/ant_demo/blob/main/img/bug_06.png?raw=true)

~~能否不抛出异常，只输出一条warning什么的。不然我觉得上层逻辑写起来心智压力也太大了，每次遍历前还得判断下目录是否存在。一般在这种情况下，目录不存在就等于目录里面是空的。~~

能否在控制台打印出异常信息，我现在必须得挂调试器才知道发生了啥事情~

**Comment by wkp850** [comment]
你可以先用 `lfs.exists()` 判断是否存在

**Reply by constly** [reply]
嗯，我在我本地加个容错。 只是对于这种情况，底层到底应不应该抛异常，可能确实要斟酌下（思考

**Reply by constly** [reply]
我回忆了下，c#之类的好像也是这样，只是在unity中搞开发时，发生这种事情后程序不会挂掉。ant现在还没完善的异常处理机制，把这种问题的后果给放大了

**Reply by wkp850** [reply]
这个看使用习惯吧，在 c++ 中，`std::filesystem::directory_iterator` 如果传不存在的目录，也会抛  `std::filesystem::filesystem_error` 异常

**Comment by actboy168** [comment]
这个库是对c++的std::filesystem的封装，所以大部分行为都会保持一致。

fs.exists/fs.is_directory不是一个轻量级的操作，如果不是必须应该尽可能减少使用。这就是为什么不会做提前检查，包括c++的std::filesystem也是这样。

搜了一下引擎对fs.pairs的使用，大多数情况都无需判断fs.is_directory，因为在上下文中已经隐含了这个条件的成立。只在需要的时候检查，或者用fs.create_directories确保它总是对的。

**Reply by constly** [reply]
感谢，学习了！
